main/tests/test_shader_lang.cpp

```diff
@@ -310,39 +310,61 @@ MainLoop *test() {
 
 	if (cmdlargs.empty()) {
 		//try editor!
-		print_line("usage: godot -test shader_lang <shader>");
+		print_line("usage: godot -test shaderlang <shader>");
 		return nullptr;
 	}
 
-	String test = cmdlargs.back()->get();
+	String code;
 
-	FileAccess *fa = FileAccess::open(test, FileAccess::READ);
+	Vector<StringName> rm;
+	rm.push_back("test_rm");
 
-	if (!fa) {
-		ERR_FAIL_V(nullptr);
-	}
+	Map<StringName, SL::FunctionInfo> dt;
+	Set<String> types;
 
-	String code;
+	if (cmdlargs.empty() || cmdlargs.back()->get() == "shaderlang") {
+#ifdef _3D_DISABLED
+		code += "shader_type canvas_item;\n";
+		code += "render_mode test_rm;\n";
+		code += "\n";
+		code += "void fragment() {\n";
+		code += "\tCOLOR = vec4(1.0);\n";
+		code += "\tdiscard;\n";
+		code += "}\n";
+#else
+		code += "shader_type spatial;\n";
+		code += "render_mode test_rm;\n";
+		code += "\n";
+		code += "void fragment() {\n";
+		code += "\tALBEDO = vec3(1.0);\n";
+		code += "\tdiscard;\n";
+		code += "}\n";
+#endif
+	} else {
+		FileAccess *fa = FileAccess::open(cmdlargs.back()->get(), FileAccess::READ);
 
-	while (true) {
-		CharType c = fa->get_8();
-		if (fa->eof_reached()) {
-			break;
+		if (!fa) {
+			ERR_FAIL_V(nullptr);
+		}
+		while (true) {
+			CharType c = fa->get_8();
+			if (fa->eof_reached()) {
+				break;
+			}
+			code += c;
 		}
-		code += c;
 	}
-
-	SL sl;
-	print_line("tokens:\n\n" + sl.token_debug(code));
-
-	Map<StringName, SL::FunctionInfo> dt;
+#ifdef _3D_DISABLED
+	types.insert("canvas_item");
+	dt["fragment"].built_ins["COLOR"] = SL::TYPE_VEC4;
+#else
+	types.insert("spatial");
 	dt["fragment"].built_ins["ALBEDO"] = SL::TYPE_VEC3;
+#endif
```

Comment:
This is wrong, that's why `canvas_item` support seems broken.
`_3D_DISABLED` is only compiled with `scons disable_3d=yes` and that removes support for 3D nodes - but when 3D support is not disabled, the `canvas_item` shaders should still be supported.

So this should be:
```suggestion
	types.insert("canvas_item");
	dt["fragment"].built_ins["COLOR"] = SL::TYPE_VEC4;
#ifndef _3D_DISABLED
	types.insert("spatial");
	dt["fragment"].built_ins["ALBEDO"] = SL::TYPE_VEC3;
#endif
```

Reply:
Fixed